File: packages/shared/types/notification.ts
/**
 * Types pour les notifications utilisateur
 * Architecture groupée logiquement pour meilleure organisation
 *
 * IMPORTANT: Le titre est construit dynamiquement côté frontend via i18n
 * à partir du `type`, `actor`, `context` et `metadata`.
 */

// =====================================================
// NOTIFICATION TYPES & ENUMS
// =====================================================

/**
 * Types de notifications supportés - ENUM complet
 */
export enum NotificationTypeEnum {
  // ===== MESSAGE EVENTS =====
  NEW_MESSAGE = 'new_message',
  MESSAGE_REPLY = 'message_reply',
  MESSAGE_EDITED = 'message_edited',
  MESSAGE_DELETED = 'message_deleted',
  MESSAGE_PINNED = 'message_pinned',
  MESSAGE_UNPINNED = 'message_unpinned',
  MESSAGE_FORWARDED = 'message_forwarded',

  // ===== CONVERSATION EVENTS =====
  NEW_CONVERSATION = 'new_conversation',
  NEW_CONVERSATION_DIRECT = 'new_conversation_direct',
  NEW_CONVERSATION_GROUP = 'new_conversation_group',
  CONVERSATION_ARCHIVED = 'conversation_archived',
  CONVERSATION_UNARCHIVED = 'conversation_unarchived',
  CONVERSATION_DELETED = 'conversation_deleted',
  CONVERSATION_SETTINGS_CHANGED = 'conversation_settings_changed',
  ADDED_TO_CONVERSATION = 'added_to_conversation',
  REMOVED_FROM_CONVERSATION = 'removed_from_conversation',
  CONVERSATION_ENCRYPTION_ENABLED = 'conversation_encryption_enabled',

  // ===== MEMBER/GROUP EVENTS =====
  MEMBER_JOINED = 'member_joined',
  MEMBER_LEFT = 'member_left',
  MEMBER_REMOVED = 'member_removed',
  MEMBER_PROMOTED = 'member_promoted',
  MEMBER_DEMOTED = 'member_demoted',
  MEMBER_ROLE_CHANGED = 'member_role_changed',

  // ===== CONTACT/FRIEND EVENTS =====
  CONTACT_REQUEST = 'contact_request',
  CONTACT_ACCEPTED = 'contact_accepted',
  CONTACT_REJECTED = 'contact_rejected',
  CONTACT_BLOCKED = 'contact_blocked',
  CONTACT_UNBLOCKED = 'contact_unblocked',
  FRIEND_REQUEST = 'friend_request',
  FRIEND_ACCEPTED = 'friend_accepted',

  // ===== INTERACTION EVENTS =====
  USER_MENTIONED = 'user_mentioned',
  MENTION = 'mention',
  MESSAGE_REACTION = 'message_reaction',
  REACTION = 'reaction',
  REPLY = 'reply',

  // ===== CALL EVENTS =====
  MISSED_CALL = 'missed_call',
  INCOMING_CALL = 'incoming_call',
  CALL_ENDED = 'call_ended',
  CALL_DECLINED = 'call_declined',
  CALL_RECORDING_READY = 'call_recording_ready',

  // ===== TRANSLATION/AUDIO EVENTS =====
  TRANSLATION_COMPLETED = 'translation_completed',
  TRANSLATION_FAILED = 'translation_failed',
  TRANSLATION_READY = 'translation_ready',
  TRANSCRIPTION_COMPLETED = 'transcription_completed',
  TRANSCRIPTION_FAILED = 'transcription_failed',
  VOICE_CLONE_READY = 'voice_clone_ready',
  VOICE_CLONE_FAILED = 'voice_clone_failed',
  AUDIO_MESSAGE_TRANSLATED = 'audio_message_translated',

  // ===== SECURITY/ACCOUNT EVENTS =====
  SECURITY_ALERT = 'security_alert',
  LOGIN_NEW_DEVICE = 'login_new_device',
  LOGIN_SUSPICIOUS = 'login_suspicious',
  PASSWORD_CHANGED = 'password_changed',
  PASSWORD_RESET_REQUESTED = 'password_reset_requested',
  EMAIL_VERIFIED = 'email_verified',
  PHONE_VERIFIED = 'phone_verified',
  TWO_FACTOR_ENABLED = 'two_factor_enabled',
  TWO_FACTOR_DISABLED = 'two_factor_disabled',
  SESSION_EXPIRED = 'session_expired',
  ACCOUNT_LOCKED = 'account_locked',
  ACCOUNT_UNLOCKED = 'account_unlocked',

  // ===== MODERATION EVENTS =====
  CONTENT_FLAGGED = 'content_flagged',
  CONTENT_REMOVED = 'content_removed',
  REPORT_SUBMITTED = 'report_submitted',
  REPORT_RESOLVED = 'report_resolved',
  WARNING_RECEIVED = 'warning_received',

  // ===== FILE/ATTACHMENT EVENTS =====
  FILE_SHARED = 'file_shared',
  FILE_UPLOAD_COMPLETED = 'file_upload_completed',
  FILE_UPLOAD_FAILED = 'file_upload_failed',
  FILE_SCAN_COMPLETED = 'file_scan_completed',

  // ===== COMMUNITY EVENTS =====
  COMMUNITY_INVITE = 'community_invite',
  COMMUNITY_JOINED = 'community_joined',
  COMMUNITY_LEFT = 'community_left',
  COMMUNITY_ANNOUNCEMENT = 'community_announcement',
  COMMUNITY_ROLE_CHANGED = 'community_role_changed',

  // ===== SYSTEM EVENTS =====
  SYSTEM = 'system',
  MAINTENANCE = 'maintenance',
  UPDATE_AVAILABLE = 'update_available',
  FEATURE_ANNOUNCEMENT = 'feature_announcement',
  TERMS_UPDATED = 'terms_updated',
  PRIVACY_UPDATED = 'privacy_updated',

  // ===== ENGAGEMENT/GAMIFICATION =====
  ACHIEVEMENT_UNLOCKED = 'achievement_unlocked',
  STREAK_MILESTONE = 'streak_milestone',
  LEVEL_UP = 'level_up',
  BADGE_EARNED = 'badge_earned',

  // ===== PAYMENT/SUBSCRIPTION =====
  SUBSCRIPTION_EXPIRING = 'subscription_expiring',
  SUBSCRIPTION_RENEWED = 'subscription_renewed',
  PAYMENT_RECEIVED = 'payment_received',
  PAYMENT_FAILED = 'payment_failed',
}

/**
 * Type union de tous les types de notifications
 */
export type NotificationType = `${NotificationTypeEnum}` | string;

/**
 * Priorité de notification
 */
export type NotificationPriority = 'low' | 'normal' | 'high' | 'urgent';

// =====================================================
// STRUCTURE GROUPÉE LOGIQUEMENT
// =====================================================

/**
 * ACTOR - Qui a déclenché la notification
 * Informations sur l'utilisateur qui a effectué l'action
 */
export interface NotificationActor {
  readonly id: string;
  readonly username: string;
  readonly displayName?: string | null;
  readonly avatar?: string | null;
}

/**
 * CONTEXT - Où c'est arrivé
 * Contexte de navigation pour la notification
 */
export interface NotificationContext {
  readonly conversationId?: string;
  readonly conversationTitle?: string;
  readonly conversationType?: 'direct' | 'group' | 'public' | 'global' | 'broadcast';
  readonly messageId?: string;
  readonly originalMessageId?: string;
  readonly callSessionId?: string;
  readonly friendRequestId?: string;
  readonly reactionId?: string;
}

/**
 * STATE - Statut de lecture
 * État de la notification
 */
export interface NotificationState {
  readonly isRead: boolean;
  readonly readAt: Date | null;
  readonly createdAt: Date;
  readonly expiresAt?: Date;
}

/**
 * DELIVERY - Suivi multi-canal
 * Statut d'envoi des notifications
 */
export interface NotificationDelivery {
  readonly emailSent: boolean;
  readonly pushSent: boolean;
}

// =====================================================
// METADATA - Type-specific data (Discriminated Unions)
// =====================================================

/**
 * Metadata de base commune à toutes les notifications
 */
interface BaseNotificationMetadata {
  readonly action?: 'view_message' | 'view_conversation' | 'join_conversation' | 'accept_or_reject_contact' | 'open_call' | 'view_details' | 'update_app' | 'none';
}

/**
 * Metadata pour new_message
 */
export interface MessageNotificationMetadata extends BaseNotificationMetadata {
  readonly messagePreview: string;
  readonly attachments?: {
    readonly count: number;
    readonly firstType: 'image' | 'video' | 'audio' | 'document' | 'text' | 'code';
    readonly firstFilename: string;
  };
  readonly action: 'view_message';
}

/**
 * Metadata pour user_mentioned
 */
export interface MentionNotificationMetadata extends BaseNotificationMetadata {
  readonly messagePreview: string;
  readonly action: 'view_message';
}

/**
 * Metadata pour message_reaction
 */
export interface ReactionNotificationMetadata extends BaseNotificationMetadata {
  readonly reactionEmoji: string;
  readonly action: 'view_message';
}

/**
 * Metadata pour missed_call / call_declined
 */
export interface CallNotificationMetadata extends BaseNotificationMetadata {
  readonly callType: 'audio' | 'video';
  readonly action: 'view_conversation';
}

/**
 * Metadata pour friend_request_received
 */
export interface FriendRequestNotificationMetadata extends BaseNotificationMetadata {
  readonly action: 'accept_or_reject_contact';
}

/**
 * Metadata pour friend_request_accepted
 */
export interface FriendAcceptedNotificationMetadata extends BaseNotificationMetadata {
  readonly action: 'view_conversation';
}

/**
 * Metadata pour user_joined_conversation / user_left_conversation
 */
export interface MemberEventNotificationMetadata extends BaseNotificationMetadata {
  readonly memberCount?: number;
  readonly isMember?: boolean;
  readonly joinMethod?: 'via_link' | 'invited';
  readonly action: 'view_conversation';
}

/**
 * Metadata pour conversation_created
 */
export interface ConversationCreatedNotificationMetadata extends BaseNotificationMetadata {
  readonly memberCount?: number;
  readonly action: 'view_conversation' | 'join_conversation';
}

/**
 * Metadata pour translation_ready
 */
export interface TranslationNotificationMetadata extends BaseNotificationMetadata {
  readonly action: 'view_message';
}

/**
 * Metadata pour system_announcement
 */
export interface SystemNotificationMetadata extends BaseNotificationMetadata {
  readonly systemType?: 'maintenance' | 'security' | 'announcement' | 'feature';
  readonly action: 'view_details' | 'update_app' | 'none';
}

/**
 * Metadata générique pour autres types
 */
export interface GenericNotificationMetadata extends BaseNotificationMetadata {
  readonly [key: string]: unknown;
}

/**
 * Union discriminée de tous les types de metadata
 */
export type NotificationMetadata =
  | MessageNotificationMetadata
  | MentionNotificationMetadata
  | ReactionNotificationMetadata
  | CallNotificationMetadata
  | FriendRequestNotificationMetadata
  | FriendAcceptedNotificationMetadata
  | MemberEventNotificationMetadata
  | ConversationCreatedNotificationMetadata
  | TranslationNotificationMetadata
  | SystemNotificationMetadata
  | GenericNotificationMetadata;

// =====================================================
// NOTIFICATION - MAIN INTERFACE
// =====================================================

/**
 * Notification - Structure groupée logiquement
 *
 * IMPORTANT: Pas de champ `title` - construit dynamiquement côté frontend via i18n
 */
export interface Notification {
  // === CORE - Identité ===
  readonly id: string;
  readonly userId: string;
  readonly type: NotificationType;
  readonly priority: NotificationPriority;

  // === CONTENT - Ce qui est affiché ===
  readonly content: string; // Message preview ou contenu principal

  // === ACTOR - Qui a déclenché (optionnel) ===
  readonly actor?: NotificationActor;

  // === CONTEXT - Où c'est arrivé ===
  readonly context: NotificationContext;

  // === METADATA - Données type-spécifiques ===
  readonly metadata: NotificationMetadata;

  // === STATE - Statut ===
  readonly state: NotificationState;

  // === DELIVERY - Suivi multi-canal ===
  readonly delivery: NotificationDelivery;
}

// =====================================================
// TYPE GUARDS
// =====================================================

/**
 * Type guard pour new_message
 */
export function isMessageNotification(n: Notification): n is Notification & { metadata: MessageNotificationMetadata } {
  return n.type === 'new_message';
}

/**
 * Type guard pour user_mentioned
 */
export function isMentionNotification(n: Notification): n is Notification & { metadata: MentionNotificationMetadata } {
  return n.type === 'user_mentioned' || n.type === 'mention';
}

/**
 * Type guard pour message_reaction
 */
export function isReactionNotification(n: Notification): n is Notification & { metadata: ReactionNotificationMetadata } {
  return n.type === 'message_reaction' || n.type === 'reaction';
}

/**
 * Type guard pour missed_call / call_declined
 */
export function isCallNotification(n: Notification): n is Notification & { metadata: CallNotificationMetadata } {
  return n.type === 'missed_call' || n.type === 'call_declined' || n.type === 'incoming_call';
}

/**
 * Type guard pour friend_request_received
 */
export function isFriendRequestNotification(n: Notification): n is Notification & { metadata: FriendRequestNotificationMetadata } {
  return n.type === 'friend_request' || n.type === 'contact_request';
}

/**
 * Type guard pour member events
 */
export function isMemberEventNotification(n: Notification): n is Notification & { metadata: MemberEventNotificationMetadata } {
  return n.type === 'member_joined' || n.type === 'member_left' || n.type === 'added_to_conversation' || n.type === 'removed_from_conversation';
}

/**
 * Type guard pour system notifications
 */
export function isSystemNotification(n: Notification): n is Notification & { metadata: SystemNotificationMetadata } {
  return n.type === 'system' || n.type === 'security_alert' || n.type === 'maintenance';
}

// =====================================================
// DTOs
// =====================================================

/**
 * DTO pour créer une notification
 */
export interface CreateNotificationDTO {
  readonly userId: string;
  readonly type: NotificationType;
  readonly priority?: NotificationPriority;
  readonly content: string;
  readonly actor?: NotificationActor;
  readonly context: NotificationContext;
  readonly metadata: NotificationMetadata;
  readonly expiresAt?: Date;
}

/**
 * DTO pour mettre à jour une notification
 */
export interface UpdateNotificationDTO {
  readonly isRead?: boolean;
  readonly readAt?: Date;
  readonly emailSent?: boolean;
  readonly pushSent?: boolean;
}

/**
 * Filtres pour rechercher des notifications
 */
export interface NotificationFilters {
  readonly userId?: string;
  readonly type?: NotificationType | 'all';
  readonly isRead?: boolean;
  readonly priority?: NotificationPriority;
  readonly conversationId?: string;
  readonly senderId?: string;
  readonly startDate?: Date;
  readonly endDate?: Date;
  readonly limit?: number;
  readonly offset?: number;
  readonly sortBy?: 'createdAt' | 'priority' | 'readAt';
  readonly sortOrder?: 'asc' | 'desc';
}

/**
 * Réponse paginée pour les notifications
 */
export interface NotificationResponse {
  readonly notifications: readonly Notification[];
  readonly pagination: {
    readonly total: number;
    readonly offset: number;
    readonly limit: number;
    readonly hasMore: boolean;
  };
  readonly unreadCount: number;
}

// =====================================================
// NOTIFICATION PREFERENCES
// =====================================================

/**
 * Préférences de notifications utilisateur
 */
export interface NotificationPreference {
  readonly id: string;
  readonly userId: string;

  // === GLOBAL SETTINGS ===
  readonly pushEnabled: boolean;
  readonly emailEnabled: boolean;
  readonly soundEnabled: boolean;

  // === PER-TYPE SETTINGS ===
  readonly newMessageEnabled: boolean;
  readonly missedCallEnabled: boolean;
  readonly systemEnabled: boolean;
  readonly conversationEnabled: boolean;
  readonly replyEnabled: boolean;
  readonly mentionEnabled: boolean;
  readonly reactionEnabled: boolean;
  readonly contactRequestEnabled: boolean;
  readonly memberJoinedEnabled: boolean;

  // === DO NOT DISTURB ===
  readonly dndEnabled: boolean;
  readonly dndStartTime?: string;
  readonly dndEndTime?: string;

  // === MUTED CONVERSATIONS ===
  readonly mutedConversations?: readonly string[];

  readonly createdAt: Date;
  readonly updatedAt: Date;
}

/**
 * DTO pour créer des préférences
 */
export interface CreateNotificationPreferenceDTO {
  readonly userId: string;
  readonly pushEnabled?: boolean;
  readonly emailEnabled?: boolean;
  readonly soundEnabled?: boolean;
  readonly newMessageEnabled?: boolean;
  readonly missedCallEnabled?: boolean;
  readonly systemEnabled?: boolean;
  readonly conversationEnabled?: boolean;
  readonly replyEnabled?: boolean;
  readonly mentionEnabled?: boolean;
  readonly reactionEnabled?: boolean;
  readonly contactRequestEnabled?: boolean;
  readonly memberJoinedEnabled?: boolean;
  readonly dndEnabled?: boolean;
  readonly dndStartTime?: string;
  readonly dndEndTime?: string;
}

/**
 * DTO pour mettre à jour des préférences
 */
export interface UpdateNotificationPreferenceDTO {
  readonly pushEnabled?: boolean;
  readonly emailEnabled?: boolean;
  readonly soundEnabled?: boolean;
  readonly newMessageEnabled?: boolean;
  readonly missedCallEnabled?: boolean;
  readonly systemEnabled?: boolean;
  readonly conversationEnabled?: boolean;
  readonly replyEnabled?: boolean;
  readonly mentionEnabled?: boolean;
  readonly reactionEnabled?: boolean;
  readonly contactRequestEnabled?: boolean;
  readonly memberJoinedEnabled?: boolean;
  readonly dndEnabled?: boolean;
  readonly dndStartTime?: string;
  readonly dndEndTime?: string;
}

// =====================================================
// UTILITIES
// =====================================================

/**
 * Vérifie si une notification est expirée
 */
export function isNotificationExpired(notification: Notification): boolean {
  if (!notification.state.expiresAt) {
    return false;
  }
  return new Date() > notification.state.expiresAt;
}

/**
 * Vérifie si une notification est non lue et valide
 */
export function isNotificationUnread(notification: Notification): boolean {
  return !notification.state.isRead && !isNotificationExpired(notification);
}

/**
 * Vérifie si le mode DND est actif
 */
export function isDNDActive(prefs: NotificationPreference): boolean {
  if (!prefs.dndEnabled) {
    return false;
  }

  if (!prefs.dndStartTime || !prefs.dndEndTime) {
    return prefs.dndEnabled;
  }

  const now = new Date();
  const currentTime = `${now.getHours().toString().padStart(2, '0')}:${now.getMinutes().toString().padStart(2, '0')}`;

  const start = prefs.dndStartTime;
  const end = prefs.dndEndTime;

  // Handle overnight DND (e.g., 22:00 - 08:00)
  if (start > end) {
    return currentTime >= start || currentTime < end;
  }

  // Normal DND (e.g., 14:00 - 16:00)
  return currentTime >= start && currentTime < end;
}

/**
 * Vérifie si un type de notification est activé
 */
export function isNotificationTypeEnabled(
  prefs: NotificationPreference,
  type: NotificationType | string
): boolean {
  switch (type) {
    case 'new_message':
      return prefs.newMessageEnabled;
    case 'missed_call':
      return prefs.missedCallEnabled;
    case 'system':
    case 'security_alert':
      return prefs.systemEnabled;
    case 'new_conversation':
      return prefs.conversationEnabled;
    case 'reply':
      return prefs.replyEnabled;
    case 'mention':
    case 'user_mentioned':
      return prefs.mentionEnabled;
    case 'reaction':
    case 'message_reaction':
      return prefs.reactionEnabled;
    case 'contact_request':
    case 'friend_request':
    case 'contact_accepted':
      return prefs.contactRequestEnabled;
    case 'member_joined':
    case 'member_left':
      return prefs.memberJoinedEnabled;
    default:
      return true;
  }
}

/**
 * Détermine si une notification doit être envoyée
 */
export function shouldSendNotification(
  prefs: NotificationPreference,
  type: NotificationType | string,
  channel: 'push' | 'email'
): boolean {
  if (channel === 'push' && !prefs.pushEnabled) {
    return false;
  }
  if (channel === 'email' && !prefs.emailEnabled) {
    return false;
  }

  if (!isNotificationTypeEnabled(prefs, type)) {
    return false;
  }

  if (type !== 'security_alert' && isDNDActive(prefs)) {
    return false;
  }

  return true;
}

/**
 * Crée les préférences par défaut
 */
export function getDefaultNotificationPreferences(userId: string): CreateNotificationPreferenceDTO {
  return {
    userId,
    pushEnabled: true,
    emailEnabled: true,
    soundEnabled: true,
    newMessageEnabled: true,
    missedCallEnabled: true,
    systemEnabled: true,
    conversationEnabled: true,
    replyEnabled: true,
    mentionEnabled: true,
    reactionEnabled: true,
    contactRequestEnabled: true,
    memberJoinedEnabled: true,
    dndEnabled: false,
  };
}

File: apps/android/core/model/src/main/kotlin/me/meeshy/sdk/model/Notification.kt
package me.meeshy.sdk.model

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

@Serializable
data class RegisterDeviceTokenRequest(
    val token: String,
    val platform: String = "ios",
    val type: String = "apns",
    val apnsEnvironment: String? = null,
)

@Serializable
data class UnregisterDeviceTokenRequest(
    val token: String,
)

@Serializable
data class RegisterDeviceTokenResponse(
    val id: String? = null,
    val type: String? = null,
    val platform: String? = null,
    val deviceName: String? = null,
    val isNew: Boolean? = null,
    val message: String? = null,
)

/** Lightweight notification preferences — port of NotificationPreferences (NotificationModels.swift). */
@Serializable
data class NotificationPreferences(
    val pushEnabled: Boolean = true,
    val messageNotifications: Boolean = true,
    val socialNotifications: Boolean = true,
    val soundEnabled: Boolean = true,
)

/** Notification type — raw values match backend strings — port of MeeshyNotificationType (NotificationModels.swift). */
@Serializable
enum class MeeshyNotificationType {
    @SerialName("new_message") NEW_MESSAGE,
    @SerialName("message_reply") MESSAGE_REPLY,
    @SerialName("message_edited") MESSAGE_EDITED,
    @SerialName("message_deleted") MESSAGE_DELETED,
    @SerialName("message_pinned") MESSAGE_PINNED,
    @SerialName("message_forwarded") MESSAGE_FORWARDED,
    @SerialName("new_conversation") NEW_CONVERSATION,
    @SerialName("new_conversation_direct") NEW_CONVERSATION_DIRECT,
    @SerialName("new_conversation_group") NEW_CONVERSATION_GROUP,
    @SerialName("added_to_conversation") ADDED_TO_CONVERSATION,
    @SerialName("removed_from_conversation") REMOVED_FROM_CONVERSATION,
    @SerialName("contact_request") CONTACT_REQUEST,
    @SerialName("contact_accepted") CONTACT_ACCEPTED,
    @SerialName("friend_request") FRIEND_REQUEST,
    @SerialName("friend_accepted") FRIEND_ACCEPTED,
    @SerialName("user_mentioned") USER_MENTIONED,
    @SerialName("mention") MENTION,
    @SerialName("message_reaction") MESSAGE_REACTION,
    @SerialName("reaction") REACTION,
    @SerialName("reply") REPLY,
    @SerialName("post_like") POST_LIKE,
    @SerialName("post_comment") POST_COMMENT,
    @SerialName("post_repost") POST_REPOST,
    @SerialName("story_reaction") STORY_REACTION,
    @SerialName("status_reaction") STATUS_REACTION,
    @SerialName("comment_like") COMMENT_LIKE,
    @SerialName("comment_reply") COMMENT_REPLY,
    @SerialName("comment_reaction") COMMENT_REACTION,
    @SerialName("story_new_comment") STORY_NEW_COMMENT,
    @SerialName("friend_story_comment") FRIEND_STORY_COMMENT,
    @SerialName("story_thread_reply") STORY_THREAD_REPLY,
    @SerialName("friend_new_story") FRIEND_NEW_STORY,
    @SerialName("friend_new_post") FRIEND_NEW_POST,
    @SerialName("friend_new_mood") FRIEND_NEW_MOOD,
    @SerialName("missed_call") MISSED_CALL,
    @SerialName("incoming_call") INCOMING_CALL,
    @SerialName("call_ended") CALL_ENDED,
    @SerialName("call_declined") CALL_DECLINED,
    @SerialName("translation_completed") TRANSLATION_COMPLETED,
    @SerialName("translation_ready") TRANSLATION_READY,
    @SerialName("transcription_completed") TRANSCRIPTION_COMPLETED,
    @SerialName("voice_clone_ready") VOICE_CLONE_READY,
    @SerialName("security_alert") SECURITY_ALERT,
    @SerialName("login_new_device") LOGIN_NEW_DEVICE,
    @SerialName("password_changed") PASSWORD_CHANGED,
    @SerialName("two_factor_enabled") TWO_FACTOR_ENABLED,
    @SerialName("two_factor_disabled") TWO_FACTOR_DISABLED,
    @SerialName("community_invite") COMMUNITY_INVITE,
    @SerialName("community_joined") COMMUNITY_JOINED,
    @SerialName("community_left") COMMUNITY_LEFT,
    @SerialName("member_joined") MEMBER_JOINED,
    @SerialName("member_left") MEMBER_LEFT,
    @SerialName("member_removed") MEMBER_REMOVED,
    @SerialName("member_promoted") MEMBER_PROMOTED,
    @SerialName("member_demoted") MEMBER_DEMOTED,
    @SerialName("member_role_changed") MEMBER_ROLE_CHANGED,
    @SerialName("system") SYSTEM,
    @SerialName("maintenance") MAINTENANCE,
    @SerialName("update_available") UPDATE_AVAILABLE,
    @SerialName("achievement_unlocked") ACHIEVEMENT_UNLOCKED,
    @SerialName("streak_milestone") STREAK_MILESTONE,
    @SerialName("badge_earned") BADGE_EARNED,

    // Legacy uppercase (backward compat)
    @SerialName("NEW_MESSAGE") LEGACY_NEW_MESSAGE,
    @SerialName("MENTION") LEGACY_MENTION,
    @SerialName("MESSAGE_REACTION") LEGACY_MESSAGE_REACTION,
    @SerialName("FRIEND_REQUEST") LEGACY_FRIEND_REQUEST,
    @SerialName("FRIEND_ACCEPTED") LEGACY_FRIEND_ACCEPTED,
    @SerialName("GROUP_INVITE") LEGACY_GROUP_INVITE,
    @SerialName("GROUP_JOINED") LEGACY_GROUP_JOINED,
    @SerialName("GROUP_LEFT") LEGACY_GROUP_LEFT,
    @SerialName("CALL_MISSED") LEGACY_CALL_MISSED,
    @SerialName("CALL_INCOMING") LEGACY_CALL_INCOMING,
    @SerialName("POST_LIKE") LEGACY_POST_LIKE,
    @SerialName("POST_COMMENT") LEGACY_POST_COMMENT,
    @SerialName("STORY_REPLY") LEGACY_STORY_REPLY,
    @SerialName("AFFILIATE_SIGNUP") LEGACY_AFFILIATE_SIGNUP,
    @SerialName("ACHIEVEMENT_UNLOCKED") LEGACY_ACHIEVEMENT_UNLOCKED,
    @SerialName("SYSTEM_ALERT") LEGACY_SYSTEM_ALERT,
    @SerialName("STATUS_UPDATE") LEGACY_STATUS_UPDATE,
    @SerialName("TRANSLATION_READY") LEGACY_TRANSLATION_READY,
}

/** Who triggered a notification — port of NotificationActor (NotificationModels.swift). */
@Serializable
data class NotificationActor(
    val id: String,
    val username: String = "",
    val displayName: String? = null,
    val avatar: String? = null,
)

/** Where a notification happened — port of NotificationContext (NotificationModels.swift). */
@Serializable
data class NotificationContext(
    val conversationId: String? = null,
    val conversationTitle: String? = null,
    val conversationType: String? = null,
    val messageId: String? = null,
    val originalMessageId: String? = null,
    val callSessionId: String? = null,
    val friendRequestId: String? = null,
    val reactionId: String? = null,
    val postId: String? = null,
    val commentId: String? = null,
)

/** Read/lifecycle state of a notification — port of NotificationState (NotificationModels.swift). */
@Serializable
data class NotificationState(
    val isRead: Boolean = false,
    val readAt: String? = null,
    val createdAt: String = "",
    val expiresAt: String? = null,
)

/** Delivery channels used for a notification — port of NotificationDelivery (NotificationModels.swift). */
@Serializable
data class NotificationDelivery(
    val emailSent: Boolean = false,
    val pushSent: Boolean = false,
)

/** Extra metadata attached to a notification — port of NotificationMetadata (NotificationModels.swift). */
@Serializable
data class NotificationMetadata(
    val messagePreview: String? = null,
    val action: String? = null,
    val reactionEmoji: String? = null,
    val callType: String? = null,
    val memberCount: Int? = null,
    val postId: String? = null,
    val commentId: String? = null,
    val commentPreview: String? = null,
    val emoji: String? = null,
    val postType: String? = null,
    val deviceName: String? = null,
    val deviceVendor: String? = null,
    val deviceOS: String? = null,
    val deviceOSVersion: String? = null,
    val deviceType: String? = null,
    val ipAddress: String? = null,
    val country: String? = null,
    val countryName: String? = null,
    val city: String? = null,
    val location: String? = null,
)

/** A notification — port of APINotification (NotificationModels.swift). */
@Serializable
data class ApiNotification(
    val id: String,
    val userId: String = "",
    val type: String = "system",
    val priority: String? = null,
    val content: String? = null,
    val actor: NotificationActor? = null,
    val context: NotificationContext? = null,
    val metadata: NotificationMetadata? = null,
    val state: NotificationState = NotificationState(),
    val delivery: NotificationDelivery? = null,
)

/** Legacy navigation data computed from a notification's context — port of NotificationData (NotificationModels.swift). */
@Serializable
data class NotificationData(
    val conversationId: String? = null,
    val messageId: String? = null,
    val postId: String? = null,
    val achievementId: String? = null,
    val callId: String? = null,
    val friendRequestId: String? = null,
    val preview: String? = null,
)

@Serializable
data class NotificationListResponse(
    val success: Boolean = false,
    val data: List<ApiNotification> = emptyList(),
    val pagination: NotificationPagination? = null,
    val unreadCount: Int? = null,
)

@Serializable
data class NotificationPagination(
    val total: Int = 0,
    val offset: Int = 0,
    val limit: Int = 0,
    val hasMore: Boolean = false,
)

@Serializable
data class UnreadCountResponse(
    val success: Boolean = false,
    val count: Int = 0,
)

@Serializable
data class MarkReadResponse(
    val success: Boolean = false,
    val count: Int? = null,
)
